source/slang/slang-preprocessor.cpp

```diff
@@ -1,198 +1,648 @@
+        ///
+    enum class State
+    {
+            /// Indicates that this conditional construct has not yet encountered a branch with a `true` condition.
+            ///
+            /// The preprocessor should skip tokens, but should keep scanning and evaluating branch conditions.
+        Before,
+
+            /// Indicates that this conditional construct is nested inside the branch with a `true` condition
+            ///
+            /// The preprocessor should not skip tokens, and should not bother evaluating subsequent branch conditions.
+        During,
+
+            /// Indicates that this conditional has laready seen the branch with a `true` condition
+            ///
+            /// The preprocessor should skip tokens, and should not bother evaluating subsequent branch conditions.
+        After,
+    };
+
+        /// The next outer conditional in the current input file, or NULL if this is the outer-most conditional.
+    Conditional* parent;
+
+        /// The token that started the conditional (e.g., an `#if` or `#ifdef`)
+    Token ifToken;
+
+        /// The `#else` directive token, if one has been seen (otherwise has `TokenType::Unknown`)
+    Token elseToken;
+
+        /// The state of the conditional
+    State state;
+};
+
+    /// An environment used for mapping macro names to their definitions during preprocessing.
+    ///
+struct Environment
 {
-        /// The macro that is busy.
-    PreprocessorMacro*  macro = nullptr;
+        /// The "outer" environment, to be used if lookup in this env fails
+    Environment* parent = NULL;
+
+        /// Macros defined in this environment
+    Dictionary<Name*, MacroDefinition*>  macros;
 
-        /// The rest of the list of busy macros.
-    BusyMacro*          next = nullptr;
+        /// Clean up the environment, releasing all macros allocated into it
+    ~Environment();
 };
 
-struct PreprocessorEnvironment
+//
+// Input Streams
+//
+
+// A fundamental action in the preprocessor is to transform a stream of
+// input tokens to produce a stream of output tokens. The term "macro expansion"
+// is used to describe two inter-related transformations of this kind:
+//
+// * Given an invocation of a macro `M`, we can "play back" the tokens in the
+//   definition of `M` to produce a stream of tokens, potentially substituting
+//   in argument values for parameters, pasting tokens, etc.
+//
+// * Given an input stream, we can scan its tokens looking for macro invocations,
+//   and upon finding them expand those invocations using the first approach
+//   outlined here.
+//
+// In practice, the second kind of expansion needs to abstract over where it
+// is reading tokens from: an input file, an existing macro invocation, etc.
+// In order to support reading from streams of tokens without knowing their
+// exact implementation, we will define an abstract base class for input
+// streams.
+
+    /// A logical stream of tokens.
+struct InputStream
```

Comment:
I *think* this is just advocating for something that can handle longer runs, but I wanted to note that we certainly can't get away with a simple two-level nesting of primary/expansion streams, because there can at times be two (or more?) expansion streams in effect at a time. This is because of the way that macro arguments get expanded before substitution, and then the result of substitution is subjected to expansion again. IIUC, you either have to support arbitrary nesting of expansion streams *or* you preprocessor needs to work by explicitly pre-expanding and buffering up tokens as part of setting up a macro invocation. The latter might often be the faster thing to do, just because it means your central operation is something like `TokenList macroExpand(TokenList tokens)` which can, as you note, often just scan until it sees a macro name.